### PR TITLE
12 kun regular ifodalar

### PR DESCRIPTION
**Pull request by ochiqai**
12-kun yozilib chiqildi. Bu regular ifodalar haqida.
**Qanday qilib tekshirish**
- Tekstdagi xatolarni tekshirish 
- Kodlardagi xatolarni tekshirish

### CODE REVIEW

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+Received: from eyewitness.mr.itd.umich.edu (eyewitness.mr.itd.umich.edu [141.211.93.142])
+	by mission.mail.umich.edu () with ESMTP id m04JoHJi019755;
+	BY eyewitness.mr.itd.umich.edu ID 477E8DF2.67B91.5278 ;
+Subject: [sakai] svn commit: r39765 - in gradebook/trunk/app: business/src/java/org/sakaiproject/tool/gradebook/business business/src/java/org/sakaiproject/tool/gradebook/business/impl ui ui/src/java/org/sakaiproject/tool/gradebook/ui/helpers/beans ui/src/java/org/sakaiproject/tool/gradebook/ui/helpers/entity ui/src/java/org/sakaiproject/tool/gradebook/ui/helpers/params ui/src/java/org/sakaiproject/tool/gradebook/ui/helpers/producers ui/src/webapp/WEB-INF ui/src/webapp/WEB-INF/bundle ui/src/webapp/content/templates
+Received: from eyewitness.mr.itd.umich.edu (eyewitness.mr.itd.umich.edu [141.211.93.142])
+	BY eyewitness.mr.itd.umich.edu ID 477D8300.AC098.32562 ;
+```
+`y..i` ga o'shashso'zlar yuqoridagi qatorlarda quyidagicha kelgan:
```
Comment (Nuriddin-Dev) [inline kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:124]: o'shashso'zlar - o'xshash so'zlar bo'lish kerak edi.

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+['<source@collab.sakaiproject.org>;']
+['apache@localhost)']
+```
+
+Yuqoridagi natija elektron adreslarni chiqarishdir. Lekin, etibor bersak
+bazilarida xar xil biz xoxlamaydigan belgilar mavjud xususan `<;>`. 
+
+> Qanday qilib, ushbu belgilarni olish tashlaymiz? 
+
+Bunda bizga tor'burchak qavs yordamga keladi - `[]` to'rtburchak qavs. Bunda
```
Comment (Nuriddin-Dev) [inline kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:237]: tor'burchak - to'rtburchak

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+biz nimani xoxlashimizni oydinlashtirishimiz mumkin. Masalan, yuqoridagini aytishimiz mumkin
+faqat kichik, yoki katta yoki sonlardan tashkil topsin deb
```
Comment (Nuriddin-Dev) [inline kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:239]: faqat kichik, yoki katta yoki sonlardan - faqat kichik yoki katta harflar yoki sonlardan.

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+
+```python
+[a-zA-Z0-9]
+```
+Ya'ni
+- a-z: a dan z gacha bo'lgan harflar bo'lishi mumkin. Qisqacha hamma kichik harflar
+- A-Z: A dan Z gacha bo'lgan harflar bo'lishi mumkin. Qisqacha hamma Katta harflar
+- 0-9: 0 dan 9 gacha bo'lgan sonlar bo'lishi mumkin
+
+ Shundan foydalin quyidagi programmani yozishimiz mumkin
```
Comment (Nuriddin-Dev) [inline kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:249]: foydalin - bu so'z tushunarsiz.

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+
+```python
+tozaroq elektron addresslarni chiqarish
```
Comment (Nuriddin-Dev) [inline kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:252]: comment belgisi qolib ketgan

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+['source@collab.sakaiproject.org']
+['apache@localhost']
+['source@collab.sakaiproject.org']
+```
+
+
+## `()` Qavs
+
+Shunday holatlar bo'ladiki. Biz nafaqat biron nimani qidiramiz
+balki birion malumotni ajratib olishimizga ham to'g'ri keladi. Shunday holatlarda, qavs (`()`) dan
```
Comment (Nuriddin-Dev) [inline kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:293]: balki birion - balki biron

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+```
+
+
+## `()` Qavs
+
+Shunday holatlar bo'ladiki. Biz nafaqat biron nimani qidiramiz
+balki birion malumotni ajratib olishimizga ham to'g'ri keladi. Shunday holatlarda, qavs (`()`) dan 
+foydalanamiz. Keling birinchi quyidagi programmani ishlataylik. Bu yerda maxsus beligilar - 
+`^X\S*: [0-9.]+` shuni anglatadiki: qator boshi `X` bilan boshlansin, va undan keyin bo'shliq bo'lmagan
+bir necha belgilarkelsin toki `:`, va undan keyin ixtoyoriy haqiqiy sonlar kelsin [0-9.]+,
```
Comment (Nuriddin-Dev) [inline kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:296]: belgilarkelsin - belgilar kelsin deb ajratilib yozilishi kerak.

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+```python
+x = 'We just received $100.00 for cokking.'
+lst = re.findall("\$.[0-9.]+", x)
+print(lst)
+```
+
+Konsolda,
+
+```commandline
+['$10.00']
```
Comment (Nuriddin-Dev) [inline kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:361]: $10.00 - $100.00 bulishi kerak.

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+ko'p uchraydi. Masalan, ko'p tekstlardan faqat o'zimizga kerak bo'ladigan elektron adresslarni olish yoki sonlarni 
+olish kabi amallarni bajarishimiz mumkin. Albatta, bu amallarni o'rgangan fayldan o'qib va undan keyin `string` ni 
+qo'llasak bo'ladi. Lekin bunday qilish ko'p vaqt va energiya ketadi. Shu sababli pythonda `regular ifodalar` deb 
+nomlangan biblioteka bor. Va biz ushbu bibliotekani qisman o'rganamiz. Bu mavzu juda katta hisoblanadi
+bu haqida xatto butun kitoblar yozilgan.
+
+
+## `import re`
+
+Pythonda regular ifoda `import re` orqali bo'ladi. `re` (ko'rinib turgandek) kengaytmasi
```
Comment (ulugbekarch) [inline kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:13]: turgandek turganidek bo'ladi

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+# ^ karet bilan ishlash, qator boshini bildiradi
+import re
+
+fayl = open("regifoda.txt", 'r')
+for qator in fayl.readlines():
+    qator = qator.rstrip()
+    if re.search("^Jumladan", qator):
+        print(qator)
+```
+
```
Comment (ulugbekarch) [inline kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:90]: Mana shu xatoni oldim ishlamadi
> Traceback (most recent call last):
>   File "D:\python\git\python\kunlar\12-kun-Regular-ifodalar\12_kunni_tekshirish.py", line 3, in <module>
>     for qator in fayl.readlines():
>   File "C:\Users\NEW AGE\AppData\Local\Programs\Python\Python39\lib\encodings\cp1251.py", line 23, in decode
>     return codecs.charmap_decode(input,self.errors,decoding_table)[0]
> UnicodeDecodeError: 'charmap' codec can't decode byte 0x98 in position 55: character maps to <undefined>
Lekin men bunday qilganimda ishladi
`fayl = open("regifoda.txt", 'r',encoding='utf8')`

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+```python
+# ^ karet bilan ishlash, qator boshini bildiradi
+import re
+
+fayl = open("regifoda.txt", 'r')
+for qator in fayl.readlines():
+    qator = qator.rstrip()
+    if re.search("^Jumladan", qator):
+        print(qator)
```
Comment (Nuriddin-Dev) [inline kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:88]: _Mana shu xatoni oldim ishlamadi._

> Traceback (most recent call last):
>   File "E:\Python\git\python\kunlar\12-kun-Regular-ifodalar\12-kun_tekshirish_n.py", line 6, in <module>
>     for qator in fayl.readlines():
>   File "C:\Users\KORD\AppData\Local\Programs\Python\Python39\lib\encodings\cp1251.py", line 23, in decode
>     return codecs.charmap_decode(input,self.errors,decoding_table)[0]
> UnicodeDecodeError: 'charmap' codec can't decode byte 0x98 in position 55: character maps to <undefined>

_Lekin men bunday qilganimda ishladi._

`fayl = open("regifoda.txt", 'r', encoding="utf8")`

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md [hunk 12]
```diff
@@ -0,0 +1,412 @@
...
+kerak bo'lsin. Unda biz quyidagicha yozamiz `y..i`, yani ikki nuqta o'rniga ixtiyoriy harflar bo'lishi
+mumkinligini bildiradi. Buni quyidagi programmada qo'llasak,
+
+```python
+#  nuqtalar bilan ishlash, nuqtalar ixtiyoriy belgiga ruhsat borligini bildiradi
+fayl = open("regifoda.txt", 'r')
+for qator in fayl.readlines():
+    qator = qator.rstrip()
+    if re.search("y..i", qator):
+        print(qator)
```
Comment (Nuriddin-Dev) [inline kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:110]: _Bunda ham mana shunday qilsam ishladi._
`fayl = open("regifoda.txt", 'r', encoding="utf8")`

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md [hunk 13]
```diff
@@ -0,0 +1,412 @@
...
+ Shundan foydalin quyidagi programmani yozishimiz mumkin
+
+```python
+tozaroq elektron addresslarni chiqarish
+fayl = open("regifoda.txt", 'r')
+for qator in fayl.readlines():
+    qator = qator.rstrip()
+    lst = re.findall('[a-zA-Z0-9]\S*@\S*[a-zA-Z]', qator)
+    if lst != []:
+        print(lst)
```
Comment (Nuriddin-Dev) [inline kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:258]: tozaroq elektron addresslarni chiqarish   - comment belgisi ishlatilmagan.
`#tozaroq elektron addresslarni chiqarish `

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md [hunk 14]
```diff
@@ -0,0 +1,412 @@
...
+['<source@collab.sakaiproject.org>;']
+['<source@collab.sakaiproject.org>;']
+['<source@collab.sakaiproject.org>;']
+['apache@localhost)']
+```
+
+Yuqoridagi natija elektron adreslarni chiqarishdir. Lekin, etibor bersak
+bazilarida xar xil biz xoxlamaydigan belgilar mavjud xususan `<;>`. 
+
+> Qanday qilib, ushbu belgilarni olish tashlaymiz?
```
Comment (Nuriddin-Dev) [inline kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:235]: olish tashlaymiz - olib tashlaymiz boladi

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md (hunk 11 above)
Comment (ulugbekarch) [reply kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:88]: Mana shu xatoni oldim ishlamadi.

Traceback (most recent call last):
File "E:\Python\git\python\kunlar\12-kun-Regular-ifodalar\12-kun_tekshirish_n.py", line 6, in
for qator in fayl.readlines():
File "C:\Users\KORD\AppData\Local\Programs\Python\Python39\lib\encodings\cp1251.py", line 23, in decode
return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x98 in position 55: character maps to

Lekin men bunday qilganimda ishladi.
`
fayl = open("regifoda.txt", 'r', encoding="utf8")`

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md (hunk 4 above)
Comment (ulugbekarch) [reply kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:249]: foydalin >>>foydalanib

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md (hunk 1 above)
Comment (ochiqai) [reply kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:124]: To'g'irlandi. Rahmat.

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md (hunk 2 above)
Comment (ochiqai) [reply kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:237]: To'g'irlandi. Rahmat.

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md (hunk 3 above)
Comment (ochiqai) [reply kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:239]: To'g'irlandi. Rahmat.

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md (hunk 4 above)
Comment (ochiqai) [reply kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:249]: To'g'irlandi. Rahmat.

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md (hunk 6 above)
Comment (ochiqai) [reply kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:293]: To'g'irlandi. Rahmat.

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md (hunk 7 above)
Comment (ochiqai) [reply kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:296]: To'g'irlandi. Rahmat.

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md (hunk 8 above)
Comment (ochiqai) [reply kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:361]: To'g'irlandi. Rahmat.

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md (hunk 9 above)
Comment (ochiqai) [reply kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:13]: To'g'irlandi. Rahmat.

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md (hunk 10 above)
Comment (ochiqai) [reply kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:90]: qayerda `open` ni qo'llangan bo'lsa hammasi o'zgartirildi.

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md (hunk 11 above)
Comment (ochiqai) [reply kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:88]: qayerda `open` ni qo'llangan bo'lsa hammasi o'zgartirildi. Yani, 
> `fayl = open("regifoda.txt", 'r',encoding='utf8')`

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md (hunk 12 above)
Comment (ochiqai) [reply kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:110]: qayerda `open` ni qo'llangan bo'lsa hammasi o'zgartirildi. Yani, 
> `fayl = open("regifoda.txt", 'r',encoding='utf8')`

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md (hunk 13 above)
Comment (ochiqai) [reply kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:258]: To'g'irlandi. Rahmat.

File: kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md (hunk 14 above)
Comment (ochiqai) [reply kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md:235]: To'g'irlandi. Rahmat.

**Pull request merged**